can/bit_timing.py

```diff
@@ -83,6 +105,83 @@ def __init__(
         if nof_samples not in (1, 3):
             raise ValueError("nof_samples must be 1 or 3")
 
+        self._sample_point = sample_point
+
+    def _calc_timings(self):
+        if self._f_clock is None:
+            raise ValueError('The f_clock is needed in order to calculate the timings')
+
+        high_sample = 0
+        low_sample = 0
+        low = (0, 0, 0, 0)
+        high = (0, 0, 0, 0)
+        res = None
+
+        for can_brp, can_sjw, can_tseg1, can_tseg2 in self._calc:
```

Comment:
I guess 4 nested for loops would be more performant, because with certain values of prescaler your wanted bitrate might not be possible at all. So you can just skip than checking all combinations of can_sjw, can_tseg1, can_tseg2 for it.

Reply:
elaborate on this... I am not understanding what you are talking about.

Reply:
You have this loop over 32768 combinations, and you end the loop at a perfect match of bitrate.

But there is also the relation
   b = self._f_clock / (can_brp * (1 + can_tseg1 + can_tseg2))

so as you check with increasing values there is always the points where
   b < self._bitrate 

So checking larger values from can_brp, can_tseq1, and can_tseq2 make no sense anymore. Then you skip testing this and any larger values for this parameter.

Reply:
I just pushed a change. Is this what you are talking about? I know that list comprehensions are the ticket in terms of speed when having to iterate. and because I moved the checks into the comprehension the size of the list is now small so it has a low memory footprint.

Reply:
There are more improvements possible using nested loops:

- In case one of the params is already fixed, you can run the for-loop for it instead of the full range just over [self._brp] (or the other ones). This saves generating a lot of combinations which you just filter out now with your first check in the loop.

- It seems can_swj is not used for any calculations in the loop. So it makes no sense to iterate over it. Just check beforehand for the already set value (self._swj) or take just 1 (your loop will not give you any other value)

Reply:
nested loops are going to be far slower then list comprehension by far. I have the number of combinations down to < 64 or so with exact bitrate matches, and for not exact bitrate matches it is around 200 or so. so there isn't much iteration that needs to be done once the list is made.

the sjw is needed for sure. it is used to calculate the btr0

Reply:
using a test script for list comprehension vs nested loops and timing how long it takes to go over all 32768 entries 10 times each. I am using the timeit module and getting an average of the 10 runs

list comprehension = 0.00925539
nested loops = 0.024373795499999993

quite a difference.